### ISSUE TITLE
Use Ansi Color Codes for ColoredConsoleTarget on Linux/.Net Core

### ISSUE THREAD

**Issue by jp7677** [NLog/NLog#1486]
Hi there,

Thanks a lot for already providing a beta for .Net Core RC2. It already works pretty cool!

I would love to use the ColoredConsoleTarget on Linux. Unfortunately the output is on my Linux box identical with the normal ConsoleTarget, thus no colors. I guess because Console.ForegrondColor and friends are quite different on Linux, of even not implemented on those platforms. Looking at Microsoft's own Aspnet Core Logger they make a distinction between Windows and Non-windows. On Non-Windows they use the regular System.Console class too, but use Ansi Color codes as message prefix instead of using Console.ForegrondColor. See here [https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Console/Internal/AnsiLogConsole.cs](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Console/Internal/AnsiLogConsole.cs) and [https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs).

Probably Nlog should do something similar. I guess using Ansi Color Code is also slightly faster..

Regards and thanks a lot for your work.
Jens

**Comment by 304NotModified** [comment]
Good idea, although there a some issues with the console logger in AspNet Core:

See https://github.com/aspnet/Logging/issues/281 and https://github.com/aspnet/Logging/issues/428

**Comment by jp7677** [comment]
Thank you for your response. I have played a little bit and this is my result so far [https://github.com/NLog/NLog/compare/coreCLR...jp7677:coreCLR](https://github.com/NLog/NLog/compare/coreCLR...jp7677:coreCLR). Works fine on my Linux box ;) 
Would you be interested in a pull request if I add the missing color codes and have finished the Word highlighting? Though first I need to understand whats happening in the word highlighting code..

Regarding both issues you linked to: Well, the docker boys could use the normal ConsoleTarget or wait till the docker extension in Visual Studio is fixed. The other issue about the colors on Mac is not clear to me. We could use the ansi codes on Linux only.

**Comment by jp7677** [comment]
Ok, after playing a little bit more it seems like dotnet does some different magic with my Linux terminal when running "dotnet run" compared to "dotnet <entry assembly>". It seems like the original ColoredConsoleTarget works fine when using "dotnet <entry assembly>", but not when doing "dotnet run". Although my version does some colors on "dotnet run" and works fine with "dotnet <entry assembly>". I'll research a bit more and post my results.

PS: The platform detection code in PlatformDetector isn't woking for me. Eventually you could consider this snippet instead:
`            
if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
  return RuntimeOS.Windows;
else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
  return RuntimeOS.Unix;
else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
  return RuntimeOS.Unix;
             
return RuntimeOS.Unknown;
`

**Comment by 304NotModified** [comment]
> Ok, after playing a little bit more it seems like dotnet does some different magic with my Linux terminal when running "dotnet run" compared to "dotnet [entry assembly]". It seems like the original ColoredConsoleTarget works fine when using "dotnet [entry assembly]", but not when doing "dotnet run".

That's strange. So for now keep it like this? 



> PS: The platform detection code in PlatformDetector isn't woking for me, Microsoft.Extensions.PlatformAbstractions.PlatformServices.Default?.Runtime?.OperatingSystem return "fedora" in my case

AFAIK the  `RuntimeInformation` is newer and better? I think it's wise to move to RuntimeInformation then. Can you send a PR for it?

**Comment by jp7677** [comment]
See https://github.com/NLog/NLog/pull/1488.

**Comment by jp7677** [comment]
Regarding the colors, yes, see this screenshot made with simple Console.Write statements. 
![screenshot from 2016-05-29 20-03-14](https://cloud.githubusercontent.com/assets/17219293/15635315/a13afafe-25db-11e6-9661-41d025231e12.png)
It seems like ansi escape codes and using Console.ForegroundColor/BackgroundColor do work fine when running dotnet [assembly]. One remark, using ansi escape codes seems somewhat better with backgroundcolors and line endings, this is unfortunately not visible in the screenshot.
Although when using "dotnet run" with Console.ForegroundColor/BackgroundColor I see no colors at all and with ansi escape codes I see just the non-bright/normal colors but not the bright ones (see [https://en.wikipedia.org/wiki/ANSI_escape_code#Colors](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors)).

To summarize, using ansi escape codes are certainly an improvement, but still not perfect.

**Comment by jp7677** [comment]
What would you prefer? Keep the current version or should I give a shot to get the word highlighting working with escape codes too so that it could end up in a PR?

**Comment by 304NotModified** [comment]
If you think it's an improvement,  then please send a PR!  :+1:

**Comment by jp7677** [comment]
I'm working on it ;)

**Comment by 304NotModified** [comment]
:)

**Comment by jp7677** [comment]
Please see https://github.com/NLog/NLog/pull/1495. 
I gave it a try, i'm curious about your feedback.

**Comment by jp7677** [comment]
Please note that the original coorizer works fine now using dotnet core 1.0 RTM. I still think that my implementation is more efficient, but from a functional point of view its not really really needed anymore. Though at least for Net Core on Linux. Don't know about other target platforms.

**Comment by 304NotModified** [comment]
Thanks for the info.  If this isn't really adding functionality, it's I guess unwise to merge it (because of maintenance and added complexity) . What do you think?

**Comment by jp7677** [comment]
Well, not sure. The output from both colorizers do look identical. Performance could be reason to merge. Running:
```
System.Diagnostics.Stopwatch watch = new System.Diagnostics.Stopwatch();
watch.Start(); 
for (var i = 0; i <= 100000; i++)
{
  logger.Trace("Do some trace logging....");
  logger.Debug("Do some debug logging....");
  logger.Info("Do some info logging....");
  logger.Warn("Do some warn logging....");
  logger.Error("Do some error logging....");
  logger.Fatal("Do some fatal logging....");
}
watch.Stop();
logger.Info("Total: {0}.", watch.ElapsedMilliseconds);
```
with the following nlog.config
```
<target name="console" xsi:type="ColoredConsole" layout="${longdate} ${pad:padding=5:inner=-${level}-} ${message}"
            useDefaultRowHighlightingRules="false" >
  <highlight-row condition="level==LogLevel.Trace" foregroundColor="Gray" />
  <highlight-row condition="level==LogLevel.Debug" foregroundColor="DarkGray" />
  <highlight-word text="-Info-"  foregroundColor="DarkGreen" />
  <highlight-word text="-Warn-"  foregroundColor="DarkYellow" />
  <highlight-word text="-Error-"  foregroundColor="DarkMagenta" />
  <highlight-word text="-Fatal-"  foregroundColor="DarkRed" backgroundColor="White"/>
</target>
```

Gives me the following results:
Original implementation:
Total: 25287.
Total: 25560.
Total: 27182.

Using AnsiColorizer:
Total: 18946.
Total: 18208.
Total: 17576.

Furthermore the AnsiColorizer is nicely unit tested. But obviously I'm a little bit biased with this question :).

**Comment by jp7677** [comment]
Hi there, any idea how to proceed?

**Comment by jp7677** [comment]
Looking at this thread https://github.com/dotnet/cli/issues/1977, seems like the Windows Command Prompt knows about ansi colors in windows 10 1511+. Seems like it is not that easy to detect if you got a windows prompt at that level, but eventually this will get easier in the future so that the unix way of doing things could get the standard way to go. Having the ansi color implementation already in the field for unix based platforms would give us some valuable feedback before the windows people are switching to it. (Not that I'm currently aware of any problems in the AnsiColorizer ;) )

**Comment by 304NotModified** [comment]
Hey, too bad I haven't found time to look into it. Because it doesn't fix an issue, it had also low prio

**Comment by jp7677** [comment]
No worries and completely understandable, please take your time. Sorry for pushing a little bit, just curious what you are going to do ;).

**Comment by 304NotModified** [comment]
:+1: thanks

**Comment by jp7677** [comment]
I had moved to SeriLog some time ago (that fits my target somewhat better, net core on Linux), so there is no demand from my side anymore.

**Comment by snakefoot** [comment]
Created #3018 for resolving this in the next release NLog ver. 4.6
